### PR TITLE
[wip][apil] change overflow format to Alert.

### PR DESCRIPTION
**Pull request by registergoofy**
* Change the overflow format (SignaOcurence) to alert
* align all leakybucket code to use Alert for Overflow
* replace formatOverflow naming by NewAlert
* update pkg/types accordingly
* get rid of the use of outputs (do not delete it yet, because it may still be useful
* get rid of the use of cwapi calls that used outputs

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -1,6 +1,9 @@
+9fans.net/go v0.0.0-20181112161441-237454027057 h1:OcHlKWkAMJEF1ndWLGxp5dnJQkYM/YImUOvsBoz6h5E=
```
Comment (buixor) [inline go.sum:1]: slightly unrelated note : how do we want to handle the go.sum ?

File: pkg/leakybucket/bucket.go [hunk 2]
```diff
@@ -218,41 +220,41 @@ func LeakRoutine(l *Leaky) {
 		case <-l.KillSwitch:
 			close(l.Signal)
 			l.logger.Debugf("Bucket externally killed, return")
-			l.AllOut <- types.Event{Overflow: types.SignalOccurence{MapKey: l.Mapkey}, Type: types.OVFLW}
+			l.AllOut <- types.Event{Type: types.OVFLW}
```
Comment (buixor) [inline pkg/leakybucket/bucket.go:223]: if we don't return the MapKey, how can the bucket be properly deleted from map ?

File: pkg/leakybucket/manager_load.go [hunk 3]
```diff
@@ -160,6 +161,9 @@ func LoadBuckets(files []string, dataFolder string) ([]BucketFactory, chan types
 			g.Filename = filepath.Clean(f)
 			g.BucketName = seed.Generate()
 			g.ret = response
+			if g.Scope == "" {
+				g.Scope = "IP"
```
Comment (buixor) [inline pkg/leakybucket/manager_load.go:165]: do we want to maintain a list of "native" scopes (ie. ip/range/..) in a iota/enum of some sorts ? (avoid typos and make scope-changes easier)

File: pkg/leakybucket/overflows.go [hunk 4]
```diff
@@ -1,135 +1,107 @@
...
+				l.logger.Errorf("Declared range %s of %s can't be parsed", v, src.Ip.String())
+			} else if ipNet != nil {
+				src.Range = *ipNet
+				l.logger.Tracef("Valid range from %s : %s", src.Ip.String(), src.Range.String())
+			}
+		}
+	}
+	if l.Scope != "IP" {
+		src.Scope = l.Scope
+		if _, ok := evt.Meta[l.Scope]; ok {
```
Comment (buixor) [inline pkg/leakybucket/overflows.go:44]: I don't think we want to force the scope to be strictly equal to the map key to use, as the user writing the scenario might not have control of the parser that is generating data being processed. For example, `scope` and `scope_source_field` (should the later one even be an expr ?) could be two different fields of the configuration?

File: pkg/leakybucket/overflows.go [hunk 5]
```diff
@@ -1,135 +1,107 @@
...
-		am = fmt.Sprintf("%d IPs", len(sig.Sources))
-	} else if len(sig.Sources) == 1 {
-		if sig.Source != nil {
-			am = sig.Source.Ip.String()
-		} else {
-			am = "??"
-		}
+	//Management of Alert.Message
+	if len(alert.Sources) > 1 {
+		am = fmt.Sprintf("%d Sources on scope %s", len(alert.Sources))
```
Comment (buixor) [inline pkg/leakybucket/overflows.go:98]: format string hasn't enough params

File: pkg/leakybucket/overflows.go [hunk 6]
```diff
@@ -1,135 +1,107 @@
...
+	} else if len(alert.Sources) == 1 {
+
```
Comment (buixor) [inline pkg/leakybucket/overflows.go:100]: we might want to here display the source/source_scope

**Review by buixor — CHANGES_REQUESTED** [review]
wip

File: go.sum (hunk 1 above)
Comment (registergoofy) [reply go.sum:1]: I guess we'll ave to see how it's done by other people...

File: pkg/leakybucket/blackhole.go [hunk 7]
```diff
@@ -34,8 +34,8 @@ func NewBlackhole(g *BucketFactory) (*Blackhole, error) {
 	}, nil
 }
 
-func (bl *Blackhole) OnBucketOverflow(b *BucketFactory) func(*Leaky, types.SignalOccurence, *Queue) (types.SignalOccurence, *Queue) {
-	return func(l *Leaky, s types.SignalOccurence, q *Queue) (types.SignalOccurence, *Queue) {
+func (bl *Blackhole) OnBucketOverflow(b *BucketFactory) func(*Leaky, types.Alert, *Queue) (types.Alert, *Queue) {
+	return func(l *Leaky, s types.Alert, q *Queue) (types.Alert, *Queue) {
```
Comment (AlteredCoder) [inline pkg/leakybucket/blackhole.go:38]: can you please rename the ones letter variable by more understandable name please ?

File: pkg/leakybucket/bucket.go [hunk 8]
```diff
@@ -148,6 +149,7 @@ func FromFactory(g BucketFactory) *Leaky {
 		Leakspeed:    g.leakspeed,
 		BucketConfig: &g,
```
Comment (AlteredCoder) [inline pkg/leakybucket/bucket.go:150]: can we rename this variable `g` please ?

File: pkg/leakybucket/bucket.go [hunk 9]
```diff
@@ -218,41 +220,41 @@ func LeakRoutine(l *Leaky) {
...
 			ofw := l.Queue
-			sig := types.SignalOccurence{MapKey: l.Mapkey}
+			alert := types.Alert{Mapkey: l.Mapkey}
 
 			if l.timedOverflow {
 				BucketsOverflow.With(prometheus.Labels{"name": l.Name}).Inc()
 
-				sig = FormatOverflow(l, ofw)
+				alert = NewAlert(l, ofw)
 				for _, f := range l.BucketConfig.processors {
```
Comment (AlteredCoder) [inline pkg/leakybucket/bucket.go:256]: can we rename this `f` variable please ?

File: pkg/leakybucket/manager_load.go [hunk 10]
```diff
@@ -160,6 +161,9 @@ func LoadBuckets(files []string, dataFolder string) ([]BucketFactory, chan types
 			g.Filename = filepath.Clean(f)
 			g.BucketName = seed.Generate()
 			g.ret = response
+			if g.Scope == "" {
```
Comment (AlteredCoder) [inline pkg/leakybucket/manager_load.go:164]: can we rename this `g` variable please ?

File: pkg/leakybucket/overflow_filter.go [hunk 11]
```diff
@@ -28,16 +28,16 @@ func NewOverflowFilter(g *BucketFactory) (*OverflowFilter, error) {
...
 		g.logger.Errorf("Unable to compile filter : %v", err)
 		return nil, fmt.Errorf("unable to compile filter : %v", err)
 	}
 	return &u, nil
 }
 
-func (u *OverflowFilter) OnBucketOverflow(Bucket *BucketFactory) func(*Leaky, types.SignalOccurence, *Queue) (types.SignalOccurence, *Queue) {
-	return func(l *Leaky, s types.SignalOccurence, q *Queue) (types.SignalOccurence, *Queue) {
+func (u *OverflowFilter) OnBucketOverflow(Bucket *BucketFactory) func(*Leaky, types.Alert, *Queue) (types.Alert, *Queue) {
+	return func(l *Leaky, s types.Alert, q *Queue) (types.Alert, *Queue) {
```
Comment (AlteredCoder) [inline pkg/leakybucket/overflow_filter.go:40]: can you please rename the ones letter variable by more understandable name please ?

File: pkg/leakybucket/overflows.go (hunk 5 above)
Comment (AlteredCoder) [inline pkg/leakybucket/overflows.go:98]: can we rename `am` please

File: pkg/leakybucket/processor.go [hunk 12]
```diff
@@ -21,8 +21,8 @@ func (d *DumbProcessor) OnBucketPour(b *BucketFactory) func(types.Event, *Leaky)
 	}
 }
 
-func (d *DumbProcessor) OnBucketOverflow(b *BucketFactory) func(*Leaky, types.SignalOccurence, *Queue) (types.SignalOccurence, *Queue) {
-	return func(l *Leaky, s types.SignalOccurence, q *Queue) (types.SignalOccurence, *Queue) {
+func (d *DumbProcessor) OnBucketOverflow(b *BucketFactory) func(*Leaky, types.Alert, *Queue) (types.Alert, *Queue) {
+	return func(l *Leaky, s types.Alert, q *Queue) (types.Alert, *Queue) {
 		return s, q
```
Comment (AlteredCoder) [inline pkg/leakybucket/processor.go:26]: can you please rename the ones letter variable by more understandable name please ?

File: pkg/leakybucket/uniq.go [hunk 13]
```diff
@@ -40,8 +40,8 @@ func (u *Uniq) OnBucketPour(Bucket *BucketFactory) func(types.Event, *Leaky) *ty
 	}
 }
 
-func (u *Uniq) OnBucketOverflow(Bucket *BucketFactory) func(*Leaky, types.SignalOccurence, *Queue) (types.SignalOccurence, *Queue) {
-	return func(l *Leaky, sig types.SignalOccurence, queue *Queue) (types.SignalOccurence, *Queue) {
+func (u *Uniq) OnBucketOverflow(Bucket *BucketFactory) func(*Leaky, types.Alert, *Queue) (types.Alert, *Queue) {
+	return func(l *Leaky, sig types.Alert, queue *Queue) (types.Alert, *Queue) {
```
Comment (AlteredCoder) [inline pkg/leakybucket/uniq.go:44]: can you please rename the ones letter variable by more understandable name please ?

File: pkg/leakybucket/bucket.go (hunk 2 above)
Comment (registergoofy) [reply pkg/leakybucket/bucket.go:223]: Sure you're right. And by the way, that wasn't done for some buckets...

File: cmd/crowdsec/output.go [hunk 14]
```diff
@@ -55,19 +48,16 @@ LOOP:
...
+			if event.Overflow.Scenario == "" && event.Overflow.Mapkey != "" {
 				//log.Infof("Deleting expired entry %s", event.Overflow.MapKey)
-				buckets.Bucket_map.Delete(event.Overflow.MapKey)
+				buckets.Bucket_map.Delete(event.Overflow.Mapkey)
 			} else {
-				/*let's handle output profiles */
-				if err := output.ProcessOutput(event.Overflow, outputProfiles); err != nil {
-					log.Warningf("Error while processing overflow/output : %s", err)
-				}
+				log.Warningf("overflow : %+v", event)
```
[the remark this answers is not in the document]
Comment (buixor) [reply cmd/crowdsec/output.go:60]: as we don't have outputs yet, this must be considered as debug

**Review by buixor — APPROVED** [review]

**Pull request merged**